File: backend/personal-finance-tracker-spring-boot/src/main/java/com/petruth/personal_finance_tracker/service/BudgetService.java
package com.petruth.personal_finance_tracker.service;

import com.petruth.personal_finance_tracker.dto.BudgetDTO;
import com.petruth.personal_finance_tracker.dto.BudgetWithSpending;
import com.petruth.personal_finance_tracker.entity.Budget;

import java.util.List;

public interface BudgetService {
    BudgetDTO createBudget(BudgetDTO budgetDTO);
    List<BudgetDTO> findByUserId(Long userId);
    Budget findById(Long id);
    BudgetDTO updateBudget(Long id, BudgetDTO budgetDTO);
    void deleteById(Long id);

    // ✅ New methods for spending calculation
    BudgetWithSpending calculateBudgetWithSpending(Budget budget);
    List<BudgetWithSpending> getAllBudgetsWithSpending(Long userId);
}

File: frontend/angular-personal-finance-tracker/src/app/services/budget.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, forkJoin, map } from 'rxjs';
import { BudgetDto } from '../common/budget-dto';
import { environment } from '../../environments/environment';
import { BudgetWithSpending } from '../common/budget-with-spending';

@Injectable({
  providedIn: 'root'
})
export class BudgetService {
  private budgetUrl = `${environment.apiUrl}/budgets`;
  private userUrl = `${environment.apiUrl}/users`;

  constructor(private http: HttpClient) {}

  createBudget(budget: BudgetDto): Observable<BudgetDto> {
    return this.http.post<BudgetDto>(this.budgetUrl, budget);
  }

  updateBudget(id: number, budget: BudgetDto): Observable<void> {
    return this.http.put<void>(`${this.budgetUrl}/${id}`, budget);
  }

  deleteBudget(id: number): Observable<string> {
    return this.http.delete(`${this.budgetUrl}/${id}`,{ responseType: 'text' });
  }

  // ✅ New method: Get budgets with spending data
  getBudgetsWithSpending(userId: number): Observable<BudgetWithSpending[]> {
    return this.http.get<BudgetWithSpending[]>(`${this.userUrl}/${userId}/budgets`, {
      withCredentials: true
    });
  }
}
